x/gov/keeper.go

```diff
@@ -0,0 +1,409 @@
+	proposal := keeper.GetProposal(ctx, proposalID)
+	if proposal == nil {
+		return ErrUnknownProposal(keeper.codespace, proposalID)
+	}
+	if proposal.GetStatus() != StatusVotingPeriod {
+		return ErrInactiveProposal(keeper.codespace, proposalID)
+	}
+
+	if option != OptionYes && option != OptionAbstain && option != OptionNo && option != OptionNoWithVeto {
+		return ErrInvalidVote(keeper.codespace, VoteOptionToString(option))
+	}
+
+	vote := Vote{
+		ProposalID: proposalID,
+		Voter:      voterAddr,
+		Option:     option,
+	}
+	keeper.setVote(ctx, proposalID, voterAddr, vote)
+
+	return nil
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetVote(ctx sdk.Context, proposalID int64, voterAddr sdk.Address) (Vote, bool) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get(KeyVote(proposalID, voterAddr))
+	if bz == nil {
+		return Vote{}, false
+	}
+	var vote Vote
+	keeper.cdc.MustUnmarshalBinary(bz, &vote)
+	return vote, true
+}
+
+func (keeper Keeper) setVote(ctx sdk.Context, proposalID int64, voterAddr sdk.Address, vote Vote) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinary(vote)
+	store.Set(KeyVote(proposalID, voterAddr), bz)
+}
+
+// Gets all the votes on a specific proposal
+func (keeper Keeper) GetVotes(ctx sdk.Context, proposalID int64) sdk.Iterator {
+	store := ctx.KVStore(keeper.storeKey)
+	return sdk.KVStorePrefixIterator(store, KeyVotesSubspace(proposalID))
+}
+
+func (keeper Keeper) deleteVote(ctx sdk.Context, proposalID int64, voterAddr sdk.Address) {
+	store := ctx.KVStore(keeper.storeKey)
+	store.Delete(KeyVote(proposalID, voterAddr))
+}
+
+// =====================================================
+// Deposits
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) GetDeposit(ctx sdk.Context, proposalID int64, depositerAddr sdk.Address) (Deposit, bool) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get(KeyDeposit(proposalID, depositerAddr))
+	if bz == nil {
+		return Deposit{}, false
+	}
+	var deposit Deposit
+	keeper.cdc.MustUnmarshalBinary(bz, &deposit)
+	return deposit, true
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) setDeposit(ctx sdk.Context, proposalID int64, depositerAddr sdk.Address, deposit Deposit) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinary(deposit)
+	store.Set(KeyDeposit(proposalID, depositerAddr), bz)
+}
+
+// Gets the vote of a specific voter on a specific proposal
+func (keeper Keeper) AddDeposit(ctx sdk.Context, proposalID int64, depositerAddr sdk.Address, depositAmount sdk.Coins) (sdk.Error, bool) {
+	// Checks to see if proposal exists
+	proposal := keeper.GetProposal(ctx, proposalID)
+	if proposal == nil {
+		return ErrUnknownProposal(keeper.codespace, proposalID), false
+	}
+
+	// Check if proposal is still depositable
+	if (proposal.GetStatus() != StatusDepositPeriod) && (proposal.GetStatus() != StatusVotingPeriod) {
```

Comment:
Can we check by inclusion instead of exclusion? Safer; that way if we add a status later this is still correct.

Reply:
It basically is checking by inclusion.  It throws if not one of those two statuses.

Reply:
Ah sorry, indeed.